File: avoidFloodInCity.go
package main

import (
	"fmt"
	"sort"
)

func avoidFlood(rains []int) []int {
	n := len(rains)
	ans := make([]int, n)
	full := make(map[int]int)
	dryDays := []int{}

	for i, lake := range rains {
		if lake == 0 {
			dryDays = append(dryDays, i)
			ans[i] = 1 // placeholder
		} else {
			if prev, ok := full[lake]; ok {
				j := sort.Search(len(dryDays), func(x int) bool { return dryDays[x] > prev })
				if j == len(dryDays) {
					return []int{}
				}
				ans[dryDays[j]] = lake
				dryDays = append(dryDays[:j], dryDays[j+1:]...)
			}
			full[lake] = i
			ans[i] = -1
		}
	}
	return ans
}

/*
func main() {
	fmt.Println(avoidFlood([]int{1, 2, 3, 4}))        // [-1,-1,-1,-1]
	fmt.Println(avoidFlood([]int{1, 2, 0, 0, 2, 1}))  // [-1,-1,2,1,-1,-1] or similar
	fmt.Println(avoidFlood([]int{1, 2, 0, 1, 2}))     // []
}
*/

File: avoidFloodInCity.py
from bisect import bisect_right, insort
from typing import List

class Solution:
    def avoidFlood(self, rains: List[int]) -> List[int]:
        n = len(rains)
        full = {}        # lake -> last rain day
        dry_days = []    # sorted list of indices where rains[i] == 0
        ans = [-1] * n

        for i, lake in enumerate(rains):
            if lake == 0:
                insort(dry_days, i)
                ans[i] = 1  # placeholder; will adjust later
            else:
                if lake in full:
                    prev = full[lake]
                    j = bisect_right(dry_days, prev)
                    if j == len(dry_days):
                        return []  # no dry day available
                    dry_day = dry_days[j]
                    ans[dry_day] = lake
                    dry_days.pop(j)
                full[lake] = i
                ans[i] = -1

        return ans
